File: app/src/main/java/ru/speinmerk/weather/fragment/CityList/CityListAdapter.java
package ru.speinmerk.weather.fragment.CityList;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import ru.speinmerk.weather.R;
import ru.speinmerk.weather.model.City;

public class CityListAdapter extends RecyclerView.Adapter<CityListAdapter.ViewHolder> {

    private ArrayList<City> cities = new ArrayList<>();
    private OnSelectedCityListener listener;

    CityListAdapter(OnSelectedCityListener listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.adapter_city_list, parent,false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(cities.get(position));
    }

    @Override
    public int getItemCount() {
        return cities.size();
    }

    public void setData(List<City> cities) {
        this.cities.clear();
        this.cities.addAll(cities);
        notifyDataSetChanged();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        TextView textView;

        ViewHolder(View itemView) {
            super(itemView);
            textView = (TextView) itemView;
        }

        void bind(City city) {
            textView.setText(city.name);
            textView.setOnClickListener(v -> listener.onSelectedCity(city));
        }
    }

    interface OnSelectedCityListener {
        void onSelectedCity(City city);
    }
}

File: app/src/main/java/ru/speinmerk/weather/adapter/CityListAdapter.kt
package ru.speinmerk.weather.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import java.util.ArrayList
import ru.speinmerk.weather.R
import ru.speinmerk.weather.model.City

class CityListAdapter internal constructor(private val listener: OnSelectedCityListener) : RecyclerView.Adapter<CityListAdapter.ViewHolder>() {

    private val cities = ArrayList<City>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.adapter_city_list, parent, false)
        return ViewHolder(view as TextView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(cities[position])
    }

    override fun getItemCount(): Int {
        return cities.size
    }

    fun setData(cities: List<City>) {
        this.cities.clear()
        this.cities.addAll(cities)
        notifyDataSetChanged()
    }

    inner class ViewHolder(private val textView: TextView) : RecyclerView.ViewHolder(textView) {
        fun bind(city: City) {
            textView.text = city.name
            textView.setOnClickListener { v -> listener.onSelectedCity(city) }
        }
    }

    internal interface OnSelectedCityListener {
        fun onSelectedCity(city: City)
    }
}
